HPCC-14929 Save DESDL state to local file

rpastrana opened the pull request:
- Attempts to load DESDL Binding from local file if dali unavailable
- Attempts to load ESDL Definition from local file if dali unavailable
- Stores ESDL to local file when state changed based on dali

Signed-off-by: rpastrana rodrigo.pastrana@lexisnexis.com

rpastrana committed:
HPCC-14929 Save DESDL state to local file

- Attempts to load DESDL Binding from local file if dali unavailable
- Attempts to load ESDL Definition from local file if dali unavailable
- Stores ESDL to local file when state changed based on dali

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@afishbeck please review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
     {
-        DBGLOG("Unable to connect to ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
-        return NULL;
+        conn.set(querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT));
```
this block contains pre-exiting logic, now encapsulated in a try/catch

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
+}
 
-    DBGLOG("ESDL Binding: Fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+IPropertyTree * fetchESDLBindingFromStateFile(const char *process, const char *bindingName, const char * stateFileName)
```
New function to load ESDL binding (configuration) from state store

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
-    IPropertyTree *esdlBindings = conn->queryRoot();
-    if (!esdlBindings)
-       throw MakeStringException(-1, "Unable to open ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
+    const char * restoredBindingTS = esdlState->queryProp("@StateSaveTimems");
```
Unused, considered adding to log.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
+    return esdlBinding.getLink();
+}
+
+void saveState(const char * esdlDefinition, IPropertyTree * esdlBinding, const char * stateFileFullPath)
```
Static version of the saveState logic, callers are static. Added wrapper method to EsdlBinding class as well.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -851,28 +949,44 @@ EsdlBindingImpl::EsdlBindingImpl()
     {
-        DBGLOG("ESDL Binding %s is subscribing to all /ESDL/Bindings/Binding dali changes", binding);
-        //Since it seems we cannot subscribe to a specific /ESDL/Bindings/Binding, all bindings have to subscribe to the /ESDL/Bindings branch
-        m_pBindingSubscription.clear();
```
Moved subscription to Dali changes after successfully loading ESDL binding/definitions

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -851,28 +949,44 @@ EsdlBindingImpl::EsdlBindingImpl()
-        //Since it seems we cannot subscribe to a specific /ESDL/Bindings/Binding, all bindings have to subscribe to the /ESDL/Bindings branch
-        m_pBindingSubscription.clear();
-        m_pBindingSubscription.setown( new CESDLBindingSubscription(this) );
+        char currentDirectory[_MAX_DIR];
```
Borrowed concept of storing state files in currentdirectory from legacy code, I'm not crazy about this approach. Do we have an adequate location for this type of data?

afishbeck replied:
Current directory may be best.  Other option would be to 1. find the current component dir using the directory settings (there are routines somewhere to manipulate that).  2. use something like /var/HPCCSystems/hpcc-data/esdl/instance 

Could check how roxie decides where to save DALI based state.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -851,28 +949,44 @@ EsdlBindingImpl::EsdlBindingImpl()
-        m_pDefinitionSubscription.clear();
-        m_pDefinitionSubscription.setown( new CESDLDefinitionSubscription(this) );
+        m_esdlStateFilesLocation.setf("%s%cDESDL_STATE",currentDirectory, PATHSEPCHAR);
+        recursiveCreateDirectory(m_esdlStateFilesLocation.str());
```
One state file per independent DESDL based binding

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
+#endif
+
+    Owned<IPropertyTree> serviceESDLDef = createPTreeFromXMLString(dESDLState.str(), ipt_caseInsensitive);
+    saveXML(stateFileFullPath, serviceESDLDef);
```
In case you crash while updating, I would do something like 1. write to a file with a unique name modifier. 2. rename when file is written.

If you crash and restart without dali available during #1 (could be caused by external or other thread), then you will reload from the prior cache.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -896,14 +1010,42 @@ EsdlBindingImpl::EsdlBindingImpl(IPropertyTree* cfg,
+    try
+    {
+        //rodrigo is this the best way to fetch the ESDL def for a given service???
+        Owned<IEsdlDefObjectIterator> it = m_esdl->getDependencies(m_espServiceName.str(), "", 0, NULL, 0);
```
Hmmm.  Should we reduce our m_esdl to only these objects in the first place?  If we are doing the dependencies gathering right we shouldn't need anything else, right?

afishbeck replied:
Ok, so yeah, if we changed the way loadDefinition worked in the firstplace so we only kept a super-set of the dependencies we could just add a IEsdlDefinition::toXML(StringBuffer &) otherwise we have to do something like what is here.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -851,28 +949,44 @@ EsdlBindingImpl::EsdlBindingImpl()
+        recursiveCreateDirectory(m_esdlStateFilesLocation.str());
 
-        m_esdlBndCfg.set(fetchESDLBindingFromDali(process, binding));
+        m_esdlStateFilesLocation.appendf("%c%s-%s.xml", PATHSEPCHAR, m_processName.str(), m_bindingName.str());
```
Good this should make the name unique even if running multiple ESDL bindings on the same process.

afishbeck commented:
@rpastrana other than comments looks good.  Should make doing the detach from roxie part pretty straight forward.  Is that planned for 6.0?

rpastrana commented:
@afishbeck there's currently no plan to add a new option to detach DESDL from Roxie, but yes it's possible to get into 6.0 after these changes are finalized, I'll open a new jira.

rpastrana commented:
https://track.hpccsystems.com/browse/HPCC-15037

afishbeck commented:
@rpastrana sorry I meant DALI not Roxie.  You can close that Jira.

rpastrana commented:
Tony, I knew you meant DALI... not sure why it was written as ROXIE... I can close but we want a command to detach from DALI, correct? I can rename the JIRA.

rpastrana commented:
@afishbeck please review implementation of new feature requested.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@rpastrana this is actually quite a bit more complicated then what I was suggesting.  I think that would be a 3 line change.

You used to have:

> Owned<IPropertyTree> serviceESDLDef = createPTreeFromXMLString(dESDLState.str(), ipt_caseInsensitive);

> saveXML(stateFileFullPath, serviceESDLDef);


I was suggesting change to:

> Owned<IPropertyTree> serviceESDLDef = createPTreeFromXMLString(dESDLState.str(), ipt_caseInsensitive);

> StringBuffer tempname;

> makeTempCopyName(tmpname, stateFileFullPath);
> saveXML(tmpname, serviceESDLDef);
> renameFile(tmpname, stateFileFullPath);

makeTempCopyNamie is in jfile, you'd have to figure out the best way to do renameFile.

But the idea is just to be safe in the case where ESP creashes during saveXML by using a unique temp name just for that brief window of time.

rpastrana commented:
you just wanted to make sure the write to file succeed?

afishbeck commented:
I think the logic is more complicated than we need.  All I wanted was to make sure we don't end up with a corrupted file when we also can't attach to DALI.  There's no recovering from that without DALI.

rpastrana committed:
HPCC-14929 Code review changes

- Adds new jfile function to rename file
- Save state to temp file, and rename

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@afishbeck please review

afishbeck commented:
@rpastrana looks fine

rpastrana commented:
@richardkchapman this PR is ready to be merged.
